pinot-controller/src/main/java/com/linkedin/pinot/controller/helix/core/PinotHelixResourceManager.java

```diff
@@ -2092,131 +2090,21 @@ public PinotResourceManagerResponse toggleInstance(String instanceName, boolean
   }
 
   @Nonnull
-  public ZNRecord rebalanceTable(final String rawTableName, boolean dryRun, TableType tableType) {
-    TableConfig tableConfig = getTableConfig(rawTableName, tableType);
-    int targetNumReplicas;
-    if (tableType == TableType.REALTIME) {
-      String replicasString = tableConfig.getValidationConfig().getReplicasPerPartition();
-      try {
-        targetNumReplicas = Integer.parseInt(replicasString);
-      } catch (Exception e) {
-        throw new RuntimeException("Invalid value for replicasPerPartition:'" + replicasString + "'", e);
-      }
-    } else {
-      targetNumReplicas = Integer.parseInt(tableConfig.getValidationConfig().getReplication());
-    }
-    final String tenantName = tableConfig.getTenantConfig().getServer();
-    final String tableNameWithType = TableNameBuilder.forType(tableType).tableNameWithType(rawTableName);
+  public ZNRecord rebalanceTable(final String rawTableName, TableType tableType, RebalanceUserParams rebalanceUserParams) {
 
-    IdealState newIdealState;
-    if (dryRun) {
-      IdealState idealState = _helixAdmin.getResourceIdealState(_helixClusterName, tableNameWithType);
-      newIdealState = getRebalancedIdealState(idealState, targetNumReplicas, tableNameWithType, tenantName);
-    } else {
-      newIdealState = rebalanceResource(targetNumReplicas, tableNameWithType, tenantName);
-    }
+    TableConfig tableConfig = getTableConfig(rawTableName, tableType);
+    String tableNameWithType = tableConfig.getTableName();
+    IdealState idealState = _helixAdmin.getResourceIdealState(_helixClusterName, tableNameWithType);
+
+    RebalanceSegmentStrategy rebalanceSegmentsStrategy =
+        RebalanceSegmentStrategyFactory.getInstance().getRebalanceSegmentsStrategy(tableConfig);
+    PartitionAssignment newPartitionAssignment =
+        rebalanceSegmentsStrategy.rebalancePartitionAssignment(idealState, tableConfig, rebalanceUserParams);
+    IdealState newIdealState = rebalanceSegmentsStrategy.rebalanceIdealState(idealState, tableConfig,
```

Comment:
TableConfig could be passed and used from the constructor. However, if someone were to create an object once, and then call rebalance multiple times, any change made to the table config in between will be ignored. So I think it is better if the TableConfig comes in with every call to rebalance, so that it will reflect the latest change requested